### PR TITLE
Adding error codes to docker/runconfig package.

### PR DESCRIPTION
**Pull request by brahmaroutu**
Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

### CODE REVIEW

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -3297,7 +3297,7 @@ func (s *DockerSuite) TestRunContainerNetModeWithExposePort(c *check.C) {
 	}
 
 	out, _, err = dockerCmdWithError("run", "--expose", "5000", "--net=container:parent", "busybox")
-	if err == nil || !strings.Contains(out, "Conflicting options: --expose and the network mode (--expose)") {
+	if err == nil || !strings.Contains(out, "Conflicting options: --expose and the network mode (--net)") {
```
Comment (duglin) [inline integration-cli/docker_cli_run_test.go]: Why this change?  And, are you not able to use the GetErrorMessage(...) trick you use in the previous test file?

File: errors/runconfig.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import (
+	"net/http"
+
+	"github.com/docker/distribution/registry/api/errcode"
+)
+
+var (
+	// ErrorCodeInvalidMacAddress is generated when invalid Mac Address is provided.
+	ErrorCodeInvalidMacAddress = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "INVALIDMACADDRESS",
```
Comment (tiborvass) [inline errors/runconfig.go]: @duglin why are these values spelled like this? just curious.

File: integration-cli/docker_cli_run_test.go (hunk 1 above)
Comment (brahmaroutu) [inline integration-cli/docker_cli_run_test.go]: @duglin I think the original code meant --net instead. I corrected accordingly. It should not be --expose. Nothing to do with error codes

File: errors/runconfig.go (hunk 2 above)
Comment (brahmaroutu) [inline errors/runconfig.go]: @tiborvass do you mean something like INVALID-MAC-ADDRESS to be more readable? I like to have string constants in one single case to be easy to remember(no Camel-case)

Comment (tiborvass) [inline errors/runconfig.go]: yeah or with underscores, but I see we already have it like that in errors/ so it's out of scope for this PR, was just curious.

Comment (brahmaroutu) [inline errors/runconfig.go]: not too late to fix this. I am also having trouble to read this well. @duglin?

Comment (duglin) [inline errors/runconfig.go]: Was following the pattern that @stevvooe defined.  @stevvooe any comment?

Personally, I think having underscores (_) would be ok. I think the only constraint I would enforce is that the ID and the ErrorCode variable match (escape possibly for case).

File: integration-cli/docker_cli_run_test.go (hunk 1 above)
Comment (duglin) [inline integration-cli/docker_cli_run_test.go]: gotcha - ok thanks

Comment (duglin) [inline integration-cli/docker_cli_run_test.go]: Can you not use the errorcode message text instead of a static string?

File: errors/runconfig.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		Description:    "The specified Mac Address is invalid",
+		HTTPStatusCode: http.StatusNotFound,
+	})
+
+	// ErrorCodeInvalidMemorySwappiness is generated when invalid Memory Swappiness value is provided.
+	ErrorCodeInvalidMemorySwappiness = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "INVALIDMEMORYSWAPPINESS",
+		Message:        "Invalid value: %d. Valid memory swappiness range is 0-100",
+		Description:    "The specified Memory swappiness is invalid",
+		HTTPStatusCode: http.StatusNotFound,
```
Comment (cpuguy83) [inline errors/runconfig.go]: Should be status 422, which unfortunately doesn't have a var defined in the http package.

**Comment by cpuguy83** [comment]
These invalid instructions should be status 422, I think.

**Comment by runcom** [comment]
I'd say they should be just 400 bad request instead cause you're providing an invalid value, would this work?

**Comment by duglin** [comment]
This is why we should just keep everything at 500, like they are today, so that we can convert to use the new structure w/o getting into a possible bike-shedding discussion around what's the proper status code - which we will discuss in a follow-on phase  :-)

**Comment by brahmaroutu** [comment]
True, it can be 400 or 500 (generic). any thoughts?

**Comment by cpuguy83** [comment]
400 = the server can't read the request
422 = the request is readable, just invalid semantically

File: integration-cli/docker_cli_run_test.go (hunk 1 above)
Comment (brahmaroutu) [inline integration-cli/docker_cli_run_test.go]: changed the static strings to error codes.

**Comment by brahmaroutu** [comment]
@cpuguy83, @runcom   falling back to 500 until phase II.

File: errors/runconfig.go (hunk 2 above)
Comment (thaJeztah) [inline errors/runconfig.go]: just had the same; they're kinda hard to read without underscores. Briefly discussed it with @duglin on IRC

ping @stevvooe could you provide your opinion on this; would adding underscores be an option?

Comment (stevvooe) [inline errors/runconfig.go]: The patterns is clearly shown in registry/api/v2/errors.go in the distribution project. They should be namespaced and have underscores.

Comment (stevvooe) [inline errors/runconfig.go]: For example, `INVALIDMEMORYSWAPPINESS` should be `RUNCONFIG_INVALID_MEMORY_SWAPPINESS`. Either way, these error codes are way to detailed for use over an API. They don't provide any method to group the errors by how to handle them.

**Comment by stevvooe** [comment]
Again, as I said with #15069, this is *not* the correct usage of the `errcode` package. The error codes are meant for defining clear http errors across API boundaries. They should not be integrated into the daemon unless within an http response handler. Furthermore, these should not be integrated into the UX as we are over-coupling the HTTP APIs to the UI. I'm going to start removing features supporting it, as they are making the error path complex and hard to understand. I'll try to find some time to correctly document how to use this package, as well. Arguably, we need to add some additional methods and tooling to map package errors to API errors, but I believe this is the job of the handler.

The main issue here is that the error space in this PR is representing what has happened rather than what to do. For example, in unix, we get an `EAGAIN` indicating we should retry an operation, rather than a condition tell us what happened and having to figure out how to resolve it. Looking at the [distribution API errors](https://godoc.org/github.com/docker/distribution/registry/api/v2#pkg-variables), we see this pattern, as well. As an example, if the client gets a `BLOB_UNKNOWN` error while uploading, it means they must upload the blob.

For this particular case, I would define two errors `RUNCONFIG_ARGUMENT_CONFLICT` and `RUNCONFIG_ARGUMENT_INVALID`. We can then use the message and detail fields to communicate which field is invalid or conflicting.

I apologize if this point is subtle, but we need to be careful to not let the error code surface area expand uncontrollably.

**Comment by duglin** [comment]
@stevvooe 
I want to make sure I understand what you're describing, so let me echo back what I think you're saying:
- in the daemon code you want to generate an error (that's at a different level than what we're generating today) and include some additional details in the message/detail fields to include which subclass of error it is.
- in the CLI code we receive the error, and then examine the message/detail fields to map the error into something more detailed.

If that's correct, there are some possible concerns with this:
- first and foremost, that is not what the Docker engine does today and the biggest goal of the current exercise is to have zero UX (and HTTP API) impact as we convert from static strings in the code to static look-up strings. This is an important goal because this conversion we're doing is not something that users of Docker should be aware of - its an internal plumbing thing (for now anyway).
- you mentioned http handlers, which I assume means that you're thinking it might be possible to do the 2nd bullet above on the daemon side (via an http handler) rather than the CLI side, leaving the CLI unaware of what's going on. And that would certainly help because I wouldn't want all possible clients to have to do the error mapping that you're describing.
- this mapping between errors is going to be pretty complex. If you look at today's error processing code in the engine, at the point where the error is generated they produce an error message and that is sent back to the CLI. We are keeping this same model, just using the errocode lookup to fetch the error string. With what you're proposing, the developer would have to, in essence, create at least two errors for every one they do today. One at the point of error detection, and one (or more) during the mapping. That's a lot more work.  And they need to then find a way to determine which subclass of error it is based on data in the details/message - and they need to keep that logic in-sync at both ends, which makes it very brittle. It won't be long before people modify one side w/o the other.
- I don't see the value within the engine of splitting the logic into two, versus just producing the final error at the time of detection - after all, it already has all of the information already.
- With respect to having errors that tell the CLI what to do vs what happened, that's really a side issue because that's more a question of what we put into the error text which we can change if needed. But, as I mentioned above, my initial goal is to have zero UX impact on this first round.

**Comment by cpuguy83** [comment]
@duglin In the case of runconfig here, I'd see it that most of these errors are really 1 errcode, something like `INVALID_OPTION`, and then supply the message with the details.

**Comment by duglin** [comment]
@cpuguy83 while we're not doing translations (i18n) now, we need to set it up so we can add it easily. What that means is that all of the static text should be in the errorCode objects at compile time, when possible, so that it can be converted to the proper language. Dynamic text is usually left for non-translatable type of substitutions - container ID, image name/ID, etc..  Now, that's not to say we should never put dynamic text into the details (or in a %s in the message), but keep in mind that it should really only be done when we can't determine the text at compile time because that'll hurt the non-english user.

For example, I did a quick scan of the errorCodes in this PR and most (if not all) know exactly what is wrong and have a very clear error message that can, and should, be translated. If we had just one `INVALID_OPTION` errorCode with these friendly error messages put into the details (by the code) then we're not really moving the ball forward from what we have today - we just wrapped a code-generated string in an errorCode and the errorCode message wouldn't be much more than just a '%s', all of the useful info is hard-coded english and sitting in the code itself, not in some look-up table.

Additionally, each errorCode has a description field which can be used to include more info about what's going on/wrong and could include things like potential fixes - basically mini-docs. If we use generic errorCodes, when we have more info and could have used more specific ones, we lose the ability to provide useful info/hints in this field.

**Comment by cpuguy83** [comment]
@duglin But the description for these are all the same, the user supplied something invalid. The message provides the details of what is invalid/why it is invalid.

The part that's moving forward are:

1. the handoff to the API layer ensuring correct responses
2. not being required for craft the error message in such a way as to get the correct status code (as is the case with that `httpError` func that was being used
3. sharing errors between client/daemon

**Comment by duglin** [comment]
@cpuguy83 let's not confuse today's description with what it can hold - today it might be a bit too terse, but that's because we're lazy devs  :-)  docs people might have more to say in there.

Your 2 is right - hopefully, everyone agrees that https://github.com/docker/docker/blob/master/api/server/server.go#L233 is bad bad bad.  But, if we use a generic error code in places where we have the ability to be more precise then we're asking for that bad code to be written.  For example, in the mapping @stevvooe describes, how do you subclass the generic errorCode into something more specific? He's suggesting that you look at the details - which sounds a like that bad code, no? The if-statements written would have to look for certain strings/data in the details to make those decisions, which means the dev generating the error needs to "craft" the details section just right to make the sub-classing work properly. All we did is move that bad code from one spot to another.

Now, if people want two fields in the errorCode, one that's unique per error message/situation, and one that's more generic (like `INVALID_OPT`) then I think that might make sense, because then we can group errors into categories, but that's also a lot more work too. But perhaps something to think about.

Your 3 is something I'd like to get to as well - I'm assuming you mean expose the error ID to the CLI. But, as I mentioned above, do we really want all possible CLI devs to have to write this mapping just to get the more specific error ID?  That's asking a lot.

**Comment by LK4D4** [comment]
@duglin @brahmaroutu can we close this?
